### ISSUE TITLE
Re-entering observeChanges on the same query yields indefinitely

### ISSUE THREAD

**Issue by kostko** [meteor/meteor#4466]
If `observeChanges` is re-entered on the same query from a callback, this will cause the fiber to yield for an indefinite amount of time, thus hanging it.

For simple cases, where `observeChanges` is called on the same query directly from a callback, Meteor will already throw the following error:

```
Exception in queued task: Error: Can't call observeChanges from an observe callback on the same query
```

But in case of nested `observeChanges` calls, which run on different queries, this error will not get triggered and the fiber will hang.

I have prepared a simple reproduction in [kostko/meteor-bug-4466](https://github.com/kostko/meteor-bug-4466).

**Comment by glasser** [comment]
You're right.  The deadlock detection code is inadequate; it only detects that trying to run the task will immediately deadlock the fiber that it's in, but not that there might be a more complex deadlock.

One could in theory try to track things in a more complex way in SynchronousQueue.  Not 100% trivial though.  Definitely would review a pull request that makes safeToRunTask smarter (but not so sensitive that it doesn't let you run a task in a queue that happens to be currently running another task and waiting on something *other* than the current fiber).

**Comment by kostko** [comment]
Also, there should probably be a warning in the documentation regarding this deadlock possibility when nesting observe calls.

**Comment by hwillson** [comment]
We'll address this through documentation - tracking in https://github.com/meteor/docs/issues/178. Closing here - thanks!
